### ISSUE TITLE
Indexed property access should behave like .property access when index is immutable and of literal type

### ISSUE THREAD

**Issue by dhruvrajvanshi** [Microsoft/TypeScript#25084]
Possibly duplicate as this is easy to come across in real world codebases

## Search Terms
indexed access, string literal

## Suggestion
TL;DR `obj["some-property"]` should behave like `obj.someProperty` when the indexed key is
a literal type and const/readonly. The const/readonly part may be harder to prove for the compiler
but a conservative start is a literal expression.

## Use Cases
Something like this
```ts
obj["some-property"] ? obj["some-property"].y : null
```
The above code errors in strict checking mode because `obj["some-property"] ?` doesn't refine the type of obj["some-property"] to remove `undefined/null`.

But apart from the slightly different syntax, this isn't really different from `obj.someProperty` and should behave the same way.

## Checklist

My suggestion meets these guidelines:
* [x] This wouldn't be a breaking change in existing TypeScript / JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. new expression-level syntax)

**Comment by mhegazy** [comment]
Duplicate of https://github.com/Microsoft/TypeScript/issues/10530
